modules/tinymce/src/core/main/ts/mode/Readonly.ts

```diff
@@ -38,23 +39,50 @@ const registerReadOnlyInputBlockers = (editor: Editor): void => {
     }
   });
 
-  editor.on('input', (e) => {
-    if (!e.isComposing && isReadOnly(editor)) {
-      const undoLevel = editor.undoManager.add();
-      if (Type.isNonNullable(undoLevel)) {
-        editor.undoManager.undo();
+  // Prevent undo level additions during composition
+  editor.on('input compositionend', (e) => {
+    if (isReadOnly(editor)) {
+      // Skip undo level addition during composition
+      if (!e.isComposing) {
+        const undoLevel = editor.undoManager.add();
+        if (Type.isNonNullable(undoLevel)) {
+          editor.undoManager.undo();
+        }
       }
     }
   });
 
-  editor.on('compositionend', () => {
+  // Set up mutation observer to detect and revert unintended changes
+  const observer = new MutationObserver((mutations) => {
     if (isReadOnly(editor)) {
-      const undoLevel = editor.undoManager.add();
-      if (Type.isNonNullable(undoLevel)) {
-        editor.undoManager.undo();
+      // Skip if we're in an undo transaction (annotation changes)
+      if (editor.undoManager.typing || editor.undoManager.data.length > 0) {
+        return;
       }
+
+      mutations.forEach((mutation) => {
+        if (mutation.type === 'characterData' || mutation.type === 'childList') {
+          // Revert the changes by restoring the previous content
+          const undoLevel = editor.undoManager.add();
+          if (Type.isNonNullable(undoLevel)) {
+            editor.undoManager.undo();
```

Comment:
If there are multiple mutations this will be quite expensive. Can't we just check if any of the mutations are what we're after and revert the content once?

Reply:
Updated to batch processing, I'm thinking to have observer at all time, it should be okay now if we doing batch processing?

Reply:
yes, batch processing is good 👍 as is only connecting the observer while composing in readonly mode 👍